lvsfunc/misc.py

```diff
@@ -469,6 +472,123 @@ def colored_clips(amount: int,
+         zoom: int = 1,
+         save_location: Optional[Union[str, pathlib.Path]] = None
+         ) -> None:
+    """
+    Writes frames as RGB24 PNG files for comparison websites or local comparison.
+
+    :param clips:         Dictionary of name:clip for all clips you want to save frames from. Accepts YUV or RGB clips only.
+                          Assumes all YUV-clips are limited range and all RGB-clips are full range.
+    :param frames:        List of frame numbers to save from each clip.
+    :param random_number: Number of random frames to save from each clip (Default: 0)
+    :param folder:        Whether or not to save clips in named folders instead of prefixing the files with the clip names (Default: ``False``\ )
+    :param dither_type:   ``dither_type`` override for ``vsutil.depth`` for clips >8-bit (Default: ``None``\ )
+    :param zoom:          Zoom factor for image output. Useful for comparison websites
+                          as zooming with a browser is not reliable and might not use point/nearest-neighbor (Default: 1)
+    :param save_location: An optional folder path to save the images/sub-folders to (will normally save in the same folder
+                          as the script is ran in) (Default: ``None``\ )
+    """
+    @contextlib.contextmanager
+    def _cd(newdir):
+        prevdir = os.getcwd()
+        os.chdir(os.path.expanduser(newdir))
+        try:
+            yield
+        finally:
+            os.chdir(prevdir)
+
+    if 0 in (clip.height for clip in clips.values()) or 0 in (clip.width for clip in clips.values()):
+        raise ValueError("save: variable-resolution clips not allowed")
+    if None in (clip.format for clip in clips.values()):
+        raise ValueError("save: variable-format clips not allowed")
+
+    if save_location is not None:
+        if (save_location := pathlib.Path(save_location)).exists():
+            os.chdir(save_location)
+
+    if len(clips) > 1:
+        if len(set(clip.height for clip in clips.values())) > 1 or len(set(clip.width for clip in clips.values())) > 1:
+            raise ValueError("save: all clips must have same dimensions")
+
+    max_frame = min(clip.num_frames for clip in clips.values()) - 1
+    if not all(f < max_frame for f in frames):
+        raise ValueError("save: specified frame numbers out of range on one or more clips")
+    if len(frames) == 0 and random_number == 0:
+        random_number = 1
+
+    if random_number := abs(random_number):
+        if random_number == 1:
+            while (rand_frame_no := random.randint(0, max_frame)) in frames:  # prevents adding the same random frame number as one specified in `frames`
+                pass
+            frames.append(rand_frame_no)
+        else:
+            random_frame_numbers = random.sample(range(max_frame), random_number)
+            while not all(i not in frames for i in random_frame_numbers):
+                random_frame_numbers = random.sample(range(max_frame), random_number)
```

Comment:
This can loop infinitely if `random_number > max_frame - len(frames)`.

Reply:
I *think* I fixed this properly in the latest series of commits, a bit more logic added to prevent this and a situation where the user requests 10 frames and 5 random ones in a 12 frame clip infinitely looping as well since they should only be able to request 2 more random ones.

Reply:
There could still be issues when the user specifies the same frame multiple times or specifies frames outside of the clip length, I believe. Those can be addressed by converting to a set and checking the min and max of that, respectively. Not sure how important that would be, but I could imagine situations where someone specified the last frame to be saved and then trims the clip.

Reply:
Yep, actually already did convert frames to a set.